Support for OpInfo-based testing for operators in JIT

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48627 Refactored OpInfo testing to support custom SampleInputs, added addmm to op_db to test
* **#47696 Support for OpInfo-based testing for operators in JIT**
* #48567 Refactored assertAutodiff test to have better error message
* #47695 Refactor existing JIT testing utils to enable new OpInfo test suite to reuse existing logic

Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

Commit:
Support for OpInfo-based testing for operators in JIT

[ghstack-poisoned]

Lilyjjo commented on test/test_ops.py:
```diff
@@ -125,13 +131,146 @@ def test_inplace_gradgrad(self, device, dtype, op):
+            inplace = op.get_inplace()
+            variants = (v for v in (method, inplace) if v is not None)
+
+            # Test eager consistency
```
I'm not sure if we need to test eager's consistency here?

mruberry replied:
It could be moved to a separate test if that's simpler.

Lilyjjo replied:
Okay, I split it out into it's own test function

Lilyjjo commented:
@eellison I haven't added the functionality to test the fusion from PR #47382 yet. I can if you want me to on this  PR or on a PR stacked on top of it

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"




[ghstack-poisoned]

eellison commented:
Hey, will get to this tomorrow. I think the changes I made in that PR can be done as a follow up - they depend on some stuff i haven't landed yet.

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"




[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"




[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
 
 
-class TestOut(TestCase):
+class TestCommon(JitCommonTestCase):
```
add comment here explaining class's purpose and why it inherits from JitCommonTestCase

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+            variants = (v for v in (method, inplace) if v is not None)
+            # Computes expected forward
+
+            expected_forward = op(sample.input, *sample.args, **sample.kwargs)
```
Add comment explaining this calls function variant and is used as gold standard

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                self.assertEqual(variant_forward, expected_forward)
+
+                # Compares variant's backward
+                if variant is not inplace and op.test_inplace_grad:
```
and -> or ?

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                    name = op.name + "_"
+                    func_type = 'inplace'
+
+                with disable_autodiff_subgraph_inlining():
```
comment explaining this context manager

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                        output = attr(*inputs[1:], **kwargs)
+                        return op.output_func(output)
+
+                    # Check scripted forward and grad
```
and gradgrad

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                                            fn, 
+                                            (sample.input,) + sample.args, 
+                                            sample.kwargs, 
+                                            no_grad=(dtype not in floating_and_complex_types()))
```
the dtype functions are confusing but I think you need this one:

https://github.com/pytorch/pytorch/blob/febc76a5c69302c7d4b481632bcbd25e5d9b76ca/torch/testing/__init__.py#L293

and then you have to add half + bfloat16 manually, unfortunately

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                                            no_grad=(dtype not in floating_and_complex_types()))
+
+                    # Check alias annotation schema for correctness (make sure inputs that aren't supposed to be modified aren't)
+                    # Op-writer TODO: make sure op supports one of the operators in below list to ensure alias annotation is checked
```
Note: only runs in float32 and int64 because dtype isn't expected to affect schema

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                    # Check autodiff of nodes for traced and scripted graphs
+                    # only need to check once per sample 
+                    if dtype is torch.float32:
+                        if IS_SANDCASTLE:
```
Comment why SANDCASTLE is special

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                    if dtype in [torch.float32, torch.int32]:
+                        check_alias_annotation(name, (sample.input,) + sample.args, sample.kwargs)
+
+                    # Check autodiff of nodes for traced and scripted graphs
```
Comment here should explain fusible vs non-fusible nodes

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -63,6 +63,10 @@ def __init__(self,
                  dtypesIfCUDA=None,  # dtypes this function is expected to work with on CUDA
                  dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
                  test_inplace_grad=True,  # whether to gradcheck and gradgradcheck the inplace variant
+                 is_autodiffed=False,  # if a op's aten::node is expected to be symbolically autodiffed
```
Can I accidentally support autodiff?

mruberry replied:
This can be addressed by improving the error message on the stacked PR, see it for a comment

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -63,6 +63,10 @@ def __init__(self,
                  dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
                  test_inplace_grad=True,  # whether to gradcheck and gradgradcheck the inplace variant
+                 is_autodiffed=False,  # if a op's aten::node is expected to be symbolically autodiffed
+                 autodiff_nonfusible_nodes=None,  # ex: ['aten::add', 'aten::mm'], default is populated to be ['aten::name']
```
Add the expected type and explain how these are used, here and below.

For example, "a list of strings with node names that should not appear in any fusion group when this operation is fused, ex: ['aten::add', 'aten::mm']. If none, defaults to ['aten::<name of operator>'].

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -91,6 +95,16 @@ def __init__(self,
         self.decorators = decorators
+        self.output_func = output_func
+
+        self.is_autodiffed = is_autodiffed
```
Add an addmm OpInfo below with one sample input to test the fusible nodes case

Lilyjjo replied:
We read the addmm method test wrong, it has two nonfusible nodes instead of fusible nodes. So adding this op wouldn't help test out the fusible node case ☹️

mruberry replied:
Doh!

Want to pick a different op?

mruberry replied:
Also - do we need an example with two nonfusible nodes to test with, too? I forget the details.

Lilyjjo replied:
So, I wasn't able to find any actual test that uses the 'fusible node case' in test_jit.py or elsewhere in the repository... I'm not sure why it was a part of the original assertAutodiffNode() function. I can take it out? @eellison Do you know if keeping this unused code would be helpful?

Lilyjjo replied:
I can add the addmm test for the sake of testing the two nonfusible nodes in OpInfo if you think it's necessary. I'm not sure really why it would be. Addmm is the only op in the original method tests that has more than one nonfusible node. Other tests in test_jit.py have more than one nonfusible node but nothing in OpInfo

mruberry replied:
It'd be nice to see all the new fields, including `autodiff_nonfusible_nodes` and `autodiff_fusible_nodes`, actually get used. It doesn't have to be part of this PR if that would cause a delay.

That's wild that are no ops testing for fusions. If that's the case this PR could mark the relevant code with a comment like, "NOT CURRENTLY USED" or "Intended to be used when..."

Lilyjjo replied:
I added notes in the assertAutoDiff code and on the OpInfo class mentioning how the fusible_node case isn't used currently. I also added addmm to the commit above this one, I had to do some refactoring on how the sample inputs were being unpacked so it turned into a larger project

eellison commented on test/test_ops.py:
```diff
@@ -125,13 +131,162 @@ def test_inplace_gradgrad(self, device, dtype, op):
+            # Test eager consistency
+            for variant in variants:
+                # Verifies that inplace operations that promote int->float fail
+                # on tensors with integer dtypes.
```
This seems like a very specific test for what is supposed to be testing general consistency.

mruberry replied:
Unfortunately this check must be performed because it determines if it's possible to perform the operation inplace at all.

eellison reviewed (commented):
looks great!

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"




[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"




[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -125,13 +131,178 @@ def test_inplace_gradgrad(self, device, dtype, op):
+                #   autodiff support. Context manager forces the graph to contain
+                #   DifferentiableGraph nodes if they are present
+                with disable_autodiff_subgraph_inlining():
+                    def fn(*inputs, **kwargs):
```
Is this function still used?

Lilyjjo replied:
Yep! in check_against_reference below

mruberry reviewed (approved):
This is awesome! This PR is well-written and precise. Updating each op requires a lot of care. I'm glad this PR is happening now instead of later when there are even more OpInfos.

I added a comment about a few lines that may be dead code.

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]

Commit:
Update on "Support for OpInfo-based testing for operators in JIT"


Differential Revision: [D25212436](https://our.internmc.facebook.com/intern/diff/D25212436)

[ghstack-poisoned]
